File: src/main/java/me/lokka30/levelledmobs/customdrops/CustomDropProcessingInfo.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.customdrops;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.UUID;

import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.managers.DebugManager;
import me.lokka30.levelledmobs.misc.DebugType;
import me.lokka30.levelledmobs.wrappers.LivingEntityWrapper;
import me.lokka30.levelledmobs.rules.CustomDropsRuleSet;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Used to store information when a custom drop item is being requested either during mob spawn in
 * for equipped items or after mob death to get the items the mob will potentially drop
 *
 * @author stumper66
 * @since 2.4.1
 */
public class CustomDropProcessingInfo {

    CustomDropProcessingInfo() {
        this.groupIDsDroppedAlready = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
        this.itemsDroppedById = new TreeMap<>();
        this.allDropInstances = new LinkedList<>();
        this.playerLevelVariableCache = new TreeMap<>();
        this.stackToItem = new LinkedList<>();
        this.debugMessages = new TreeMap<>();
    }

    public LivingEntityWrapper lmEntity;
    Player mobKiller;
    final @NotNull Map<String, Integer> playerLevelVariableCache;
    DeathCause deathCause;
    double addition;
    boolean isSpawner;
    boolean equippedOnly;
    boolean deathByFire;
    boolean wasKilledByPlayer;
    boolean doNotMultiplyDrops;
    boolean hasOverride;
    boolean hasCustomDropId;
    boolean hasEquippedItems;
    int retryNumber;
    float equippedChanceRole;
    public @Nullable GroupLimits groupLimits;
    public String customDropId;
    List<ItemStack> newDrops;
    public CustomDropInstance dropInstance;
    final @NotNull Map<String, Integer> groupIDsDroppedAlready;
    final @NotNull Map<UUID, Integer> itemsDroppedById;
    Map<Integer, List<CustomDropBase>> prioritizedDrops;
    @Nullable CustomDropsRuleSet dropRules;
    final @NotNull List<CustomDropInstance> allDropInstances;
    private final Map<DebugType, StringBuilder> debugMessages;
    public final List<Map.Entry<ItemStack, CustomDropItem>> stackToItem;

    public void itemGotDropped(final @NotNull CustomDropBase dropBase, final int amountDropped){

        if (dropBase.hasGroupId()) {
            final int count = groupIDsDroppedAlready.getOrDefault(
                    dropBase.groupId, 0) + amountDropped;

            groupIDsDroppedAlready.put(dropBase.groupId, count);
        }

        final int count = itemsDroppedById.getOrDefault(dropBase.uid, 0) + amountDropped;
        itemsDroppedById.put(dropBase.uid, count);
    }

    public int getDropItemsCountForGroup(final @NotNull CustomDropBase dropBase){
        final String useGroupId = dropBase.hasGroupId() ?
                dropBase.groupId : "default";

        return groupIDsDroppedAlready.getOrDefault(useGroupId, 0);
    }

    public int getItemsDropsById(final @NotNull CustomDropBase dropBase){
        return itemsDroppedById.getOrDefault(dropBase.uid, 0);
    }

    public int getItemsDropsByGroup(final @NotNull CustomDropBase dropBase){
        final String useGroupId = dropBase.hasGroupId() ?
                dropBase.groupId : "default";

        return groupIDsDroppedAlready.getOrDefault(useGroupId, 0);
    }

    void addDebugMessage(final String message) {
        addDebugMessage(DebugType.CUSTOM_DROPS, message);
    }

    void addDebugMessage(final DebugType debugType, final String message) {
        if (!LevelledMobs.getInstance().debugManager.isDebugTypeEnabled(debugType)){
            return;
        }

        final StringBuilder sb = this.debugMessages.computeIfAbsent(
                debugType, k -> new StringBuilder());

        if (!sb.isEmpty())
            sb.append(System.lineSeparator());

        sb.append(message);
    }

    void writeAnyDebugMessages() {
        if (this.debugMessages.isEmpty()) return;

        for (DebugType debugType : this.debugMessages.keySet()){
            DebugManager.log(debugType, lmEntity, () -> this.debugMessages.get(debugType).toString());
        }

        this.debugMessages.clear();
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/customdrops/CustomDropProcessingInfo.kt
package io.github.arcaneplugins.levelledmobs.customdrops

import java.util.TreeMap
import java.util.UUID
import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.debug.DebugManager
import io.github.arcaneplugins.levelledmobs.debug.DebugType
import io.github.arcaneplugins.levelledmobs.enums.DeathCause
import io.github.arcaneplugins.levelledmobs.enums.EquipmentClass
import io.github.arcaneplugins.levelledmobs.rules.CustomDropsRuleSet
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

/**
 * Used to store information when a custom drop item is being requested either during mob spawn in
 * for equipped items or after mob death to get the items the mob will potentially drop
 *
 * @author stumper66
 * @since 2.4.1
 */
class CustomDropProcessingInfo {
    var lmEntity: LivingEntityWrapper? = null
    var mobKiller: Player? = null
    val playerLevelVariableCache = mutableMapOf<String, Int>()
    var deathCause: DeathCause? = null
    var addition = 0f
    var isSpawner = false
    var equippedOnly = false
    var deathByFire = false
    var wasKilledByPlayer = false
    var doNotMultiplyDrops  = false
    var hasOverride = false
    var hasCustomDropId = false
    var hasEquippedItems = false
    var retryNumber = 0
    var equippedChanceRole = 0f
    var groupLimits: GroupLimits? = null
    var customDropId: String? = null
    var itemWasEquipped = false
    var equipmentClass: EquipmentClass? = null
    var newDrops: MutableList<ItemStack>? = null
    var dropInstance: CustomDropInstance? = null
    var equippedItemsInfo: EquippedItemsInfo? = null
    var overallChanceDebugMessage: String = ""
    private val groupIDsDroppedAlready: MutableMap<String, Int> = TreeMap(String.CASE_INSENSITIVE_ORDER)
    private val itemsDroppedById = mutableMapOf<UUID, Int>()
    var prioritizedDrops: MutableMap<Int, MutableList<CustomDropBase>>? = null
    var dropRules: CustomDropsRuleSet? = null
    val allDropInstances=  mutableListOf<CustomDropInstance>()
    private var debugMessages: StringBuilder? = null
    val stackToItem = mutableListOf<Map.Entry<ItemStack, CustomDropItem>>()

    fun itemGotDropped(dropBase: CustomDropBase, amountDropped: Int) {
        if (dropBase.hasGroupId) {
            val count = groupIDsDroppedAlready.getOrDefault(
                dropBase.groupId, 0
            ) + amountDropped

            groupIDsDroppedAlready[dropBase.groupId!!] = count
        }

        val count = itemsDroppedById.getOrDefault(dropBase.uid, 0) + amountDropped
        itemsDroppedById[dropBase.uid] = count

        if (equipmentClass != null && dropBase is CustomDropItem && equippedItemsInfo != null){
            EquippedItemsInfo.droppedEquipmentByClass.add(equipmentClass!!)
        }
    }

    fun getDropItemsCountForGroup(dropBase: CustomDropBase): Int {
        val useGroupId = if (dropBase.hasGroupId) dropBase.groupId else "default"

        return groupIDsDroppedAlready.getOrDefault(useGroupId, 0)
    }

    fun getItemsDropsById(dropBase: CustomDropBase): Int {
        return itemsDroppedById.getOrDefault(dropBase.uid, 0)
    }

    fun getItemsDropsByGroup(dropBase: CustomDropBase): Int {
        val useGroupId = if (dropBase.hasGroupId) dropBase.groupId else "default"

        return groupIDsDroppedAlready.getOrDefault(useGroupId, 0)
    }

    fun addDebugMessage(debugType: DebugType, message: String?) {
        if (!LevelledMobs.instance.debugManager.isDebugTypeEnabled(debugType)) {
            return
        }

        addDebugMessage(message)
    }

    fun addDebugMessage(message: String?) {
        if (this.debugMessages == null) {
            this.debugMessages = StringBuilder()
        }

        if (debugMessages!!.isNotEmpty()) {
            debugMessages!!.append(System.lineSeparator())
        }

        debugMessages!!.append(message)
    }

    fun writeAnyDebugMessages() {
        if (this.debugMessages == null || debugMessages!!.isEmpty()) {
            return
        }

        DebugManager.log(DebugType.CUSTOM_DROPS, lmEntity) { debugMessages.toString() }
        debugMessages!!.setLength(0)
    }
}